hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/HoodieCompactor.java

```diff
@@ -146,7 +146,7 @@ public HoodieData<WriteStatus> compact(
 
     boolean useFileGroupReaderBasedCompaction = context.supportsFileGroupReader()   // the engine needs to support fg reader first
         && !metaClient.isMetadataTable()
-        && config.getBooleanOrDefault(HoodieReaderConfig.FILE_GROUP_READER_ENABLED)
+        && HoodieReaderConfig.isFileGroupReaderEnabled(metaClient.getTableConfig().getTableVersion(), config)
```

Comment:
The file group reader `HoodieFileGroupReader` should be the only way of reading records from a file group eventually.  Let's keep that goal in mind and do not introduce any new fallback.

Reply:
https://github.com/apache/hudi/pull/12935